drivers/media/v4l2-core/v4l2-mem2mem.c

```diff
@@ -301,9 +301,10 @@ static void __v4l2_m2m_try_queue(struct v4l2_m2m_dev *m2m_dev,
 
 	dprintk("Trying to schedule a job for m2m_ctx: %p\n", m2m_ctx);
 
-	if (!m2m_ctx->out_q_ctx.q.streaming
-	    || !m2m_ctx->cap_q_ctx.q.streaming) {
-		dprintk("Streaming needs to be on for both queues\n");
+	if ((!m2m_ctx->out_q_ctx.q.streaming ||
+	     !m2m_ctx->cap_q_ctx.q.streaming) &&
+	    !(m2m_ctx->out_q_ctx.buffered && m2m_ctx->out_q_ctx.q.streaming)) {
```

Comment:
I just spent a minute or two writing out your condition as you described it:
```
	if (!((m2m_ctx->out_q_ctx.q.streaming &&
	       m2m_ctx->cap_q_ctx.q.streaming) ||
	     ( m2m_ctx->out_q_ctx.buffered && m2m_ctx->out_q_ctx.q.streaming))) {
```
and then using De Morgan's laws to transform it into what you have written. But I shouldn't have to - what's wrong with the literal version above?

Reply:
The OP's version is just an evolution of pre-existing (and thus assumed stable to some degree through testing/deployment) code using a minimal set of changes.

The check could also be written as

```
	if (!(m2m_ctx->out_q_ctx.q.streaming && m2m_ctx-cap_q_ctx.q.streaming) && 
	       !(m2m_ctx->out_q_ctx.buffered && m2m_ctx->out_q_ctx.q.streaming)) {
```

They are all logically equivalent but having OR inside the clause and top-level NOT in the proposed transformed version makes it a slower performing check because no short-circuiting happens inside and the result is not final until after the negation is performed at the end.

The changes suggested in c54a951 still employ a short-circuit, so they look ok to me.

Reply:
They aren't wrong, but in (especially) kernel work clarity trumps pretty much everything, especially when the compiler will likely generate the same code in both cases.

Reply:
I agree on clarity, thus my suggestion above. Of course, as more conditions get added, the clause may have to be rewritten again.

Reply:
I was just retaining the format of the original term. I'll rework as requested.